js/modules/k6/ws/ws.go

```diff
@@ -0,0 +1,416 @@
+	}
+
+	s.pingSendTimestamps[pingID] = time.Now()
+	s.pingSendCounter++
+}
+
+func (s *Socket) trackPong(pingID string) {
+	pongTimestamp := time.Now()
+
+	if _, ok := s.pingSendTimestamps[pingID]; !ok {
+		// We received a pong for a ping we didn't send; ignore
+		// (this shouldn't happen with a compliant server)
+		return
+	}
+	pingTimestamp := s.pingSendTimestamps[pingID]
+
+	s.totalPingDuration += pongTimestamp.Sub(pingTimestamp)
+	s.totalPingCount++
+}
+
+func (s *Socket) SetTimeout(fn goja.Callable, timeoutMs int) {
+	// Starts a goroutine, blocks once on the timeout and pushes the callable
+	// back to the main loop through the scheduled channel
+	go func() {
+		select {
+		case <-time.After(time.Duration(timeoutMs) * time.Millisecond):
+			s.scheduled <- fn
+
+		case <-s.done:
+			return
+		}
+	}()
+}
+
+func (s *Socket) SetInterval(fn goja.Callable, intervalMs int) {
+	// Starts a goroutine, blocks forever on the ticker and pushes the callable
+	// back to the main loop through the scheduled channel
+	go func() {
+		ticker := time.NewTicker(time.Duration(intervalMs) * time.Millisecond)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				s.scheduled <- fn
+
+			case <-s.done:
+				return
+			}
+		}
+	}()
+}
+
+func (s *Socket) Close(args ...goja.Value) {
+	code := websocket.CloseGoingAway
+	if len(args) > 0 {
+		code = int(args[0].ToInteger())
+	}
+
+	s.closeConnection(code)
+}
+
+// Attempts to close the websocket gracefully
+func (s *Socket) closeConnection(code int) error {
+
+	var err error
+	s.shutdownOnce.Do(func() {
+		rt := common.GetRuntime(s.ctx)
+
+		err := s.conn.WriteControl(websocket.CloseMessage,
+			websocket.FormatCloseMessage(code, ""),
+			time.Now().Add(writeWait),
+		)
+		if err != nil {
+			s.handleEvent("error", rt.ToValue(err))
+			// Just call the handler, we'll try to close the connection anyway
+		}
+		s.conn.Close()
+
+		// Stops the main control loop
+		close(s.done)
+	})
+
+	return err
+}
+
+// Wraps conn.ReadMessage in a channel
+func readPump(conn *websocket.Conn, readChan chan []byte, errorChan chan error) {
+	defer conn.Close()
+
+	for {
+		_, message, err := conn.ReadMessage()
+		if err != nil {
+			// Only emit the error if we didn't close the socket ourselves
+			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway) {
+				errorChan <- err
+			}
+
+			return
+		}
+
+		readChan <- message
+	}
+}
+
+func parseArgs(rt *goja.Runtime, args []goja.Value) (goja.Callable, map[string]string, http.Header, error) {
+	var callableV goja.Value
+	var paramsV goja.Value
+
+	// The params argument is optional
+	switch len(args) {
+	case 2:
+		paramsV = args[0]
+		callableV = args[1]
+	case 1:
+		paramsV = goja.Undefined()
+		callableV = args[0]
+	default:
+		return nil, nil, nil, errors.New("Invalid number of arguments to ws.connect")
+	}
+
+	// Get the callable (required)
+	var callable goja.Callable
+	var isFunc bool
+	if callable, isFunc = goja.AssertFunction(callableV); !isFunc {
```

Comment:
Right, I'm a bit more used to C aesthetics I guess. I split the assignments out of the if statements to avoid those declarations in the couple of cases you mentioned.